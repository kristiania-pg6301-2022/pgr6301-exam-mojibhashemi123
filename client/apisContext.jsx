import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";

export const APIsContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },

  async registerLogin(provider, token) {
    return await postJSON(`/api/login/${provider}`, token);
  },

  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
});
