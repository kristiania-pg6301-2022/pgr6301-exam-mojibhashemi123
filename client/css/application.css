#app{
    display: grid;
    height: 100vh;
    grid-template-rows: auto 1fr;
}

.menu-divider2{
    flex: 100%;
}

body {
    margin: 0;
}

header {
    background-color: grey;
    padding: 1em 1em;
    border: black;
    border-bottom:2px solid black;
    display: flex;
    justify-content: space-between;
    border-bottom: 1px solid black;
}

header > * {
    padding-inline: 0.5em;
}

header .menu-divider{
    flex: 1;
}

#frontpage{
    padding: 10px;
    margin: 10px;
}

#loginPage{
    position: absolute;
    width: 300px;
    height: 200px;
    z-index: 15;
    top: 50%;
    left: 50%;
    margin: -100px 0 0 -150px;
}

#loginPage h1 {
    margin-bottom: 20px;
    margin-top: 20px;
}