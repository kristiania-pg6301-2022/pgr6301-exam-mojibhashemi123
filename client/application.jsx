import { BrowserRouter, Link, Route, Routes } from "react-router-dom";
import React, { useContext } from "react";
import { FrontPage } from "./pages/frontPage";
import { LoginPage } from "./pages/loginPage";
import { APIsContext } from "./apisContext";
import { useLoading } from "./useLoading";

import "./css/application.css";
import { Profile } from "./pages/profile";
import { CreateAccount } from "./pages/createAccount";
import { ListArticle, ShowMatch } from "./pages/listArticle";
import { CreateNewArticle } from "./pages/createNewArticle";
import { UpdateArticle } from "./pages/updateArticle";

export function UserActions({ user }) {
  if (!user || Object.keys(user).length === 0) {
    return <Link to={"/login"}>Login</Link>;
  }

  if (user.google?.name) {
    return (
      <>
        <Link to={"/frontPage"}>Front Page</Link>
        <Link to={"/"}>List article</Link>
        <div className="menu-divider2" />
        <Link to={"/profile"}>{`Profile for ${user.google.name}`}</Link>
        <Link to={"/login/endsession"}>Log out</Link>
      </>
    );
  } else if (user.microsoft?.name) {
    return (
      <>
        <Link to={"/frontPage"}>Front Page</Link>
        <Link to={"/"}>List article</Link>
        <div className="menu-divider2" />
        <Link to={"/profile"}>{"Profile for " + user.microsoft.name}</Link>
        <Link to={"/login/endsession"}>Log out</Link>
      </>
    );
  } else if (user.email) {
    const name = user.email.map((u) => u.name);
    return (
      <>
        <Link to={"/frontPage"}>Front Page</Link>
        <Link to={"/"}>List article</Link>
        <div className="menu-divider2" />
        <Link to={"/profile"}>{"Profile for " + name}</Link>
        <Link to={"/login/endsession"}>Log out</Link>
      </>
    );
  }
}

export function Application() {
  const { fetchLogin } = useContext(APIsContext);

  const { data, error, loading, reload } = useLoading(fetchLogin);

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  if (loading) {
    return <div>Please wait...</div>;
  }

  return (
    <BrowserRouter>
      <header>
        <UserActions user={data?.user} />
      </header>

      <main>
        <Routes>
          <Route path={"/"} element={<ListArticle />} />
          <Route path={"/frontPage"} element={<FrontPage />} />
          <Route
            path={"/article/new"}
            element={<CreateNewArticle user={data?.user} reload={reload} />}
          />
          <Route
            path={"/article/update"}
            element={<UpdateArticle user={data?.user} reload={reload} />}
          />
          <Route path={"/ShowMatch"} element={<ShowMatch />} />

          <Route
            path={"/account/new"}
            element={<CreateAccount reload={reload} />}
          />
          <Route
            path={"/login/*"}
            element={<LoginPage config={data.config} reload={reload} />}
          />
          <Route path={"/profile"} element={<Profile user={data?.user} />} />
          <Route path={"*"} element={<h1>Not found</h1>} />
        </Routes>
      </main>
    </BrowserRouter>
  );
}
