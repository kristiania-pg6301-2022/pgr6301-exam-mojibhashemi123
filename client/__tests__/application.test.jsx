import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { APIsContext } from "../apisContext";
import { Application } from "../application";

jest.mock("../css/application.css", () => jest.fn());
jest.mock("../css/loginWithLocalUser.css", () => jest.fn());

describe("fetch users", () => {
  describe("fetchlogin component", () => {
    it("shows loading screen", () => {
      const domElement = document.createElement("div");
      ReactDOM.render(<Application />, domElement);
      expect(domElement.innerHTML).toMatchSnapshot();
    });

    it("shows error message", async () => {
      const domElement = document.createElement("div");
      await act(async () => {
        const fetchLogin = () => {
          throw new Error("Something went wrong");
        };
        ReactDOM.render(
          <APIsContext.Provider value={{ fetchLogin }}>
            <Application />
          </APIsContext.Provider>,
          domElement
        );
      });

      expect(domElement.querySelector("#error-text").innerHTML).toEqual(
        "Error: Something went wrong"
      );
      expect(domElement.innerHTML).toMatchSnapshot();
    });
  });
});
