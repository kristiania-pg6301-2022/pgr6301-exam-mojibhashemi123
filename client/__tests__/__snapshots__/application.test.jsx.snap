// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch users fetchlogin component shows error message 1`] = `"<div><h1>Error</h1><div id=\\"error-text\\">Error: Something went wrong</div></div>"`;

exports[`fetch users fetchlogin component shows loading screen 1`] = `"<div>Please wait...</div>"`;
